libtiledbsoma/test/common.cc

```diff
@@ -117,71 +132,139 @@ create_arrow_schema_and_index_columns(
+        dim_schema->name = strdup(info.name.c_str());
+        dim_schema->n_children = 0;  // leaf node
+        dim_schema->dictionary = nullptr;
+        dim_schema->release = &ArrowAdapter::release_schema;
+        index_cols_info_schema->children[i] = dim_schema;
+    }
 
     return index_cols_info_schema;
 }
 
 static std::unique_ptr<ArrowArray> _create_index_cols_info_array(
-    int64_t dim_max, bool use_current_domain) {
+    const std::vector<DimInfo>& dim_infos) {
+    int ndim = dim_infos.size();
+
     auto index_cols_info_array = std::make_unique<ArrowArray>();
     index_cols_info_array->length = 0;
     index_cols_info_array->null_count = 0;
     index_cols_info_array->offset = 0;
     index_cols_info_array->n_buffers = 0;
     index_cols_info_array->buffers = nullptr;
-    index_cols_info_array->n_children = 1;
+    index_cols_info_array->n_children = ndim;
     index_cols_info_array->release = &ArrowAdapter::release_array;
-    index_cols_info_array->children = new ArrowArray*[1];
-
-    int n = use_current_domain ? 5 : 3;
-
-    auto d0_info = index_cols_info_array->children[0] = new ArrowArray;
-    d0_info->length = n;
-    d0_info->null_count = 0;
-    d0_info->offset = 0;
-    d0_info->n_buffers = 2;
-    d0_info->release = &ArrowAdapter::release_array;
-    d0_info->buffers = new const void*[2];
-    d0_info->buffers[0] = nullptr;
-    d0_info->buffers[1] = malloc(sizeof(int64_t) * n);
-    d0_info->n_children = 0;
-
-    if (use_current_domain) {
-        // domain big; current_domain small
-        int64_t dom[] = {0, CORE_DOMAIN_MAX, 1, 0, dim_max};
-        std::memcpy((void*)d0_info->buffers[1], &dom, sizeof(int64_t) * n);
-    } else {
-        // domain small; current_domain feature not being used
-        int64_t dom[] = {0, dim_max, 1};
-        std::memcpy((void*)d0_info->buffers[1], &dom, sizeof(int64_t) * n);
+    index_cols_info_array->children = new ArrowArray*[ndim];
+
+    for (int i = 0; i < ndim; i++) {
+        const DimInfo& info = dim_infos[i];
+
+        int n = info.use_current_domain ? 5 : 3;
+
+        auto dim_array = new ArrowArray;
+        dim_array->length = n;
+        dim_array->null_count = 0;
+        dim_array->offset = 0;
+        dim_array->n_buffers = 2;
+        dim_array->release = &ArrowAdapter::release_array;
+        dim_array->buffers = new const void*[2];
+        dim_array->buffers[0] = nullptr;
+        dim_array->n_children = 0;  // leaf node
+        index_cols_info_array->children[i] = dim_array;
+
+        // The full user-level SOMA API supports many more index types.
+        // Here we support enough types to verify we've got variant-indexed
+        // SOMADataFrame objects baseline-tested in C++, then defer exhaustive
+        // loop-over-all-datatypes handling to Python and R.
+        if (info.tiledb_datatype == TILEDB_INT64) {
+            size_t nbytes = n * sizeof(int64_t);
+            dim_array->buffers[1] = malloc(nbytes);
```

Comment:
This comment has nothing to do with your changes and does not (and should not) be addressed in this PR. But in these test arrays we release the `ArrowArray` via `ArrowAdapter::release_array` which relies on `ManagedQuery` to `delete` the `ArrayBuffers`, so we will need a follow up PR to address this.

Reply:
I had only meant to highlight `dim_array->buffers[1] = malloc(nbytes);`.

Reply:
Awesome, thanks @nguyenv !! :)